File: backend/services/aggregator.py
"""Aggregation service – consumes exchange ticks, normalises and computes
average / best / latest price.  Simulates AWS Lambda aggregation."""

from __future__ import annotations

import time
from typing import Any

from .event_bus import event_bus

_latest_by_exchange: dict[str, dict[str, dict[str, Any]]] = {}


async def _on_tick(tick: dict[str, Any]) -> None:
    contract = tick["contract"]
    exchange = tick["exchange"]

    if contract not in _latest_by_exchange:
        _latest_by_exchange[contract] = {}
    _latest_by_exchange[contract][exchange] = tick

    await _normalize(contract)


async def _normalize(contract: str) -> None:
    entries = list(_latest_by_exchange.get(contract, {}).values())
    if not entries:
        return

    prices = [e["price"] for e in entries]
    avg = sum(prices) / len(prices)
    latest = max(entries, key=lambda e: e["timestamp"])

    exchange_prices: dict[str, float] = {e["exchange"]: e["price"] for e in entries}

    normalized = {
        "contract": contract,
        "averagePrice": round(avg, 2),
        "bestBid": round(min(prices), 2),
        "bestAsk": round(max(prices), 2),
        "latestPrice": latest["price"],
        "latestExchange": latest["exchange"],
        "timestamp": int(time.time() * 1000),
        "exchangePrices": exchange_prices,
    }
    await event_bus.publish("aggregated:price", normalized)


_unsub = None


def start_aggregator() -> None:
    global _unsub
    _unsub = event_bus.subscribe("exchange:tick", _on_tick)


def stop_aggregator() -> None:
    global _unsub
    if _unsub:
        _unsub()
        _unsub = None

File: src/services/aggregator.ts
import { eventBus } from './eventBus';
import type { PriceTick, NormalizedPrice, ExchangeName } from '../types';

const latestByExchange: Record<string, Record<ExchangeName, PriceTick>> = {};

function normalize(): void {
  const contracts = Object.keys(latestByExchange);
  for (const contract of contracts) {
    const exchangeData = latestByExchange[contract];
    const entries = Object.values(exchangeData);
    if (entries.length === 0) continue;

    const prices = entries.map((e) => e.price);
    const avg = prices.reduce((s, p) => s + p, 0) / prices.length;
    const latest = entries.reduce((a, b) =>
      a.timestamp > b.timestamp ? a : b
    );

    const exchangePrices: Record<ExchangeName, number> = {} as Record<ExchangeName, number>;
    for (const e of entries) {
      exchangePrices[e.exchange] = e.price;
    }

    const normalized: NormalizedPrice = {
      contract,
      averagePrice: Math.round(avg * 100) / 100,
      bestBid: Math.round(Math.min(...prices) * 100) / 100,
      bestAsk: Math.round(Math.max(...prices) * 100) / 100,
      latestPrice: latest.price,
      latestExchange: latest.exchange,
      timestamp: Date.now(),
      exchangePrices,
    };

    eventBus.publish('aggregated:price', normalized);
  }
}

export function startAggregator(): () => void {
  const unsub = eventBus.subscribe('exchange:tick', (data) => {
    const tick = data as PriceTick;
    if (!latestByExchange[tick.contract]) {
      latestByExchange[tick.contract] = {} as Record<ExchangeName, PriceTick>;
    }
    latestByExchange[tick.contract][tick.exchange] = tick;
    normalize();
  });

  return unsub;
}
